### PR TITLE
[SPARK-32143][SQL] Prevent a skewed join from producing too many partition splits

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?
Add a configuration `spark.sql.adaptive.skewJoin.maxPartitionSplits` to prevent a skewed join from producing too many partition splits.

### Why are the changes needed?
In handling skewed SortMergeJoin, when matching partitions from the left side and the right side both have skew and split too many partitions, the plan generation may take a very long time and finally OOM.

Even fallback to normal SMJ, the query cannot success either. So we should fast fail this query.

In below logs we can see that it took over 1 hour to generate the plan in AQE when handle a skewed join which produced too many splits.
```
20/06/30 12:31:26,271 INFO [HiveServer2-Background-Pool: Thread-821384] adaptive.OptimizeSkewedJoin:54 :
 20/06/30 12:31:26,299 INFO [HiveServer2-Background-Pool: Thread-821384] adaptive.OptimizeSkewedJoin:54 : Left side partition 1 (3 TB) is skewed, split it into *39150* parts.
 20/06/30 12:31:26,315 INFO [HiveServer2-Background-Pool: Thread-821384] adaptive.OptimizeSkewedJoin:54 : Right side partition 1 (11 TB) is skewed, split it into *17022* parts.
 20/06/30 12:32:24,952 INFO [HiveServer2-Background-Pool: Thread-821384] adaptive.OptimizeSkewedJoin:54 : Right side partition 8 (1 GB) is skewed, split it into 17 parts.

...
 20/06/30 13:27:25,158 INFO [HiveServer2-Background-Pool: Thread-821384] adaptive.AdaptiveSparkPlanExec:54 : Final plan: CollectLimit 1000
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add a ut

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -507,6 +507,15 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createWithDefaultString("256MB")
 
+  val SKEW_JOIN_MAX_PARTITION_SPLITS =
+    buildConf("spark.sql.adaptive.skewJoin.maxPartitionSplits")
+      .doc("The max partition number produced in a skewed join. Generate the plan with too many " +
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:512]: `The max number of partitions produced...`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 2]
```diff
@@ -236,6 +236,10 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
         } {
           leftSidePartitions += leftSidePartition
           rightSidePartitions += rightSidePartition
+          if (leftSidePartitions.length > conf.getConf(SQLConf.SKEW_JOIN_MAX_PARTITION_SPLITS)) {
+            throw new SparkException(s"Too many partition splits produced in handling data skew." +
+              s" The threshold is ${conf.getConf(SQLConf.SKEW_JOIN_MAX_PARTITION_SPLITS)}")
+          }
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:242]: Besides of this error message, should we also provide some hints like adjusting `advisoryPartitionSizeInBytes` so users can try workaround?

**Review by viirya — COMMENTED** [review]
> Even fallback to normal SMJ, the query cannot success either. So we should fast fail this query.

Why? If we skip skewed join optimization, normal join should still be able to run? Although it might not be good on query performance.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala (hunk 2 above)
Comment (LantaoJin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: How about it now?

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
cc @cloud-fan @JkSelf @maryannxue @HyukjinKwon

**Comment by cloud-fan** [comment]
Does it work to tune `ADVISORY_PARTITION_SIZE_IN_BYTES`?

**Pull request closed without merging**